Managers/UInputMan.cpp

```diff
@@ -1,2858 +1,1393 @@
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	int UInputMan::ReadProperty(std::string propName, Reader &reader) {
+		if (propName == "MouseSensitivity") {
+			reader >> m_MouseSensitivity;
+		} else if (propName == "P1Scheme") {
+			reader >> m_ControlScheme[PLAYER_ONE];
+		} else if (propName == "P2Scheme") {
+			reader >> m_ControlScheme[PLAYER_TWO];
+		} else if (propName == "P3Scheme") {
+			reader >> m_ControlScheme[PLAYER_THREE];
+		} else if (propName == "P4Scheme") {
+			reader >> m_ControlScheme[PLAYER_FOUR];
+		} else {
+			return Serializable::ReadProperty(propName, reader);
+		}
 		return 0;
+	}
 
-	// Lazy load of these.. can't load earlier
-	if (!m_apDeviceIcons[DEVICE_KEYB_ONLY])
-		GetSchemeIcon(0);
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	int UInputMan::Save(Writer &writer) const {
+		writer.NewLine(false, 2);
+		writer.NewDivider(false);
+		writer.NewLineString("// Input Mapping", false);
+		writer.NewLine(false);
+
+		writer.NewProperty("MouseSensitivity");
+		writer << m_MouseSensitivity;
+
+		writer.NewLine(false);
+		writer.NewLineString("// Input Devices:  0 = Keyboard Only, 1 = Mouse + Keyboard, 2 = Joystick One, 3 = Joystick Two, , 4 = Joystick Three, 5 = Joystick Four");
+		writer.NewLineString("// Scheme Presets: 0 = No Preset, 1 = WASD, 2 = Cursor Keys, 3 = XBox 360 Controller");
+
+		writer.NewLine(false, 2);
+		writer.NewDivider(false);
+		writer.NewLineString("// Player 1", false);
+		writer.NewLine(false);
+		writer.NewProperty("P1Scheme");
+		writer << m_ControlScheme[PLAYER_ONE];
+
+		writer.NewLine(false);
+		writer.NewDivider(false);
+		writer.NewLineString("// Player 2", false);
+		writer.NewLine(false);
+		writer.NewProperty("P2Scheme");
+		writer << m_ControlScheme[PLAYER_TWO];
+
+		writer.NewLine(false);
+		writer.NewDivider(false);
+		writer.NewLineString("// Player 3", false);
+		writer.NewLine(false);
+		writer.NewProperty("P3Scheme");
+		writer << m_ControlScheme[PLAYER_THREE];
+
+		writer.NewLine(false);
+		writer.NewDivider(false);
+		writer.NewLineString("// Player 4", false);
+		writer.NewLine(false);
+		writer.NewProperty("P4Scheme");
+		writer << m_ControlScheme[PLAYER_FOUR];
 
-	return m_apDeviceIcons[whichDevice];
-}
+		return 0;
+	}
 
-//////////////////////////////////////////////////////////////////////////////////////////
-// Method:          JoystickActive
-//////////////////////////////////////////////////////////////////////////////////////////
-// Description:     Returns true if specified joystick is active
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
-bool UInputMan::JoystickActive(int joystickNumber) const
-{
-	if (joystickNumber < 0 || joystickNumber >= MAX_PLAYERS)
-		return false;
+	void UInputMan::LoadDeviceIcons() {
```

Comment:
Lazy loading here means `I'm just gonna half ass this because I don't even`.
A device icon is requested as early as loading into main menu if you're configured to use a gamepad or if you load directly into any activity. Latest it's requested is when configuring players/teams before starting an activity.
So it doesn't hurt to do this on game boot.
I changed it because it just looks like dirty fuckery to me because we don't really do any real lazy loading.

Reply:
Well lazy loading basically means - instead of loading everything up front, just load it and store it the first time it's requested. The main benefit is you wouldn't have to have LoadingGUI all tied together with UInputMan like you do.
That said, this lazy loading version is pretty mediocre, so I hear ya there. Like, would LoadingGUI still need to include UInputMan if it didn't have to call this?

Maybe you can move the GetDeviceIcon call into main, since that already has to know about a lot of details?

Reply:
Yeah I know what it means, it's just bad here imo.
And for whatever reason we're already updating `UInputMan` during `LoadingSplashProgressReport` so it's already tied in.
Gonna need to investigate why exactly we're doing this, probably for some stupid reason and it can be yeeted.

And I guess I could load the icons during main menu create.

Reply:
Fair enough. Generally speaking I'd like to decouple these things as much as is reasonably possible. If you can do this (by moving it into main or whatever) please do, but if not it's not the end of the world.